### PR TITLE
tests/pkg_cayenne_lpp: fix inconsistent buffer values across architectures (native, boards)

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR is an attempt to have a more robust output check with the pkg_cayenne-lpp test application.
I also don't think it's the goal of the test target to verify that the generated payload is valid.
But PRs on the [upstream project](https://github.com/aabadie/cayenne-lpp) are very welcome ;)

It should be useful with #9007.

I tested it on native and nucleo-l073, the ouput are sligthly different but the test target passes.

### Issues/PRs references

#9007

### CODE REVIEW

**Comment by aabadie** [comment]
@cladmi, I know I'm disabling the verification of the generated payload, and I did that on purpose (see the initial comment of this PR)
`I also don't think it's the goal of the test target to verify that the generated payload is valid.` This verification should be part of the upstream library. But this is debatable.

**Comment by cladmi** [comment]
How can you test the library with itself ? If the library does not work it could say ok when it's not good.

However if you would like to do the test on the node and not in python (as we tend to), you should add a "_compare_buffer(&lpp, expected_1)" and not do only a "_print_buffer".

**Comment by aabadie** [comment]
> How can you test the library with itself ?

Adding unit tests upstream would help (this is something I have in mind). But it wouldn't solve the issue with rounding float values being different depending on the architecture. The upstream library is tested on CircleCI, i.e. on x86 arch.

At least, you pointed to origin of the problem (floating point precision) and adding the following (taken from [there](https://www.linuxtopia.org/online_books/an_introduction_to_gcc/gccintro_70.html):
```mk
ifeq (native, $(BOARD))
  CFLAGS += -msse2 -mfpmath=sse
endif
```
to the test application makefile gives consistent results between native and other boards. Maybe this should be added to the package makefile itself. Or do you see better solutions ?

**Comment by cladmi** [comment]
Maybe even to native in general.

**Comment by aabadie** [comment]
> Maybe even to native in general.

indeed, but I have no idea of the consequences of such a change :)

**Comment by cladmi** [comment]
> indeed, but I have no idea of the consequences of such a change :)

Yes. What I mean by this is for the moment it's ok to have a special `native` case in the test.

Setting a special floating point if cayenne is included changes the behavior of all floats of the application so maybe not that good.

Asking if this floating point configuration could be made default to be more accurate to nodes float handling is another separate topic.

**Comment by kaspar030** [comment]
> CFLAGS += -msse2 -mfpmath=sse

That won't work on ARM.

**Comment by kaspar030** [comment]
> But it wouldn't solve the issue with rounding float values being different depending on the architecture.

If the library doesn't clearly define floating point behaviour (which would IMO make it way too complex), float values can always end up different when using CPU instructions directly. Better change the tests so only limited precision is tested.

**Comment by aabadie** [comment]
> Better change the tests so only limited precision is tested.

I prefer keeping the values used in the test because they are the ones given by the [cayenne lpp documentation](https://mydevices.com/cayenne/docs_stage/lora/#lora-cayenne-low-power-payload) (this gives us a ground truth).
Since only a few values (3) in the output buffer are affected, I just adapted it with a regexp and added a comment. Tested on native, samr21-xpro and nucleo-l152

**Comment by cladmi** [comment]
I still prefer an explicit `if`, because it really explains what is on native and what is on a board.
Whereas with the regex, it gives 4 valid values instead on 2, and does not say which one to expect.

The test could also be implemented in main.c and not in Python. But I could do that in a further PR.

**Comment by aabadie** [comment]
> The test could also be implemented in main.c and not in Python.

I pushed a couple of commits that does that. The idea is to let the preprocessor decide which buffer is correct depending on the target. It's much cleaner.

**Comment by aabadie** [comment]
@cladmi, updated with the expected value being displayed when each test fails.

File: tests/pkg_cayenne-lpp/main.c [hunk 1]
```diff
@@ -19,41 +19,87 @@
...
+    puts("Test 1:");
+    puts("-------");
     cayenne_lpp_add_temperature(&lpp, 3, 27.2);
     cayenne_lpp_add_temperature(&lpp, 5, 25.5);
-    _print_buffer(&lpp);
+    _print_buffer(&lpp, "Result");
+    /* check buffer */
+    if (memcmp(lpp.buffer, test_buffer1, lpp.cursor) != 0) {
+        puts("");
+        _print_buffer(&lpp, "Expected");
```
Comment (cladmi) [inline tests/pkg_cayenne-lpp/main.c]: Nop, it prints the same one :p 

```
Result: 0167FFD7067104D2FB2E0000
Expected: 0167FFD7067104D2FB2E0000 FAILED
```

**Comment by aabadie** [comment]
@cladmi, my last commit is crap, I'll rework it. Sorry

**Comment by aabadie** [comment]
Done, thanks for reviewing and testing @cladmi !

**Pull request merged**